### PR TITLE
Test adsdriver

### PR DESCRIPTION
**Pull request by rerpha**
### Description of work

*Add your own description here*

### To test

*Which ticket does this PR fix?*

### Acceptance criteria

*List the acceptance criteria for the PR*

---

#### Code Review

- [ ] A copy of the manual has been placed on the shared drive
- [ ] Pertitent information has been stored in the [wiki](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Specific-Device-IOC)
- [ ] Does the IOC conform to IBEX standards?
    - [PV naming](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/PV-Naming).
    - [Disable records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Disable-records)
    - [Record simulation](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Record-Simulation)
    - [Finishing touches](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IOC-Finishing-Touches)
- [ ] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)?
- [ ] Do the IOC and support module conform to the new [build guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Reducing-Build-Dependencies)
- [ ] Have the changes been recorded appropriately in a PR for [release notes](https://github.com/ISISComputingGroup/IBEX/blob/master/release_notes/ReleaseNotes_Upcoming.md)?
- [ ] Is the device's flow control setting correct? [For most devices flow control should be OFF](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Flow-control).

### Functional Tests

- IOC responds correctly in:
    - [ ] Devsim mode
    - [ ] Recsim mode
    - [ ] Real device, if available
- [ ] Supplementary IOCs (`..._0n` where `n>1`) run correctly
- [ ] Log files do not report undefined macros (serach for `macLib: macro` to find instances of `macLib: macro [macro name] is undefined...`

### Final steps

- [ ] Update the IOC submodule in the main EPICS repo. See [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [ ] Reviewer has merged the associated PR for the [release notes](https://github.com/ISISComputingGroup/IBEX/blob/master/release_notes/ReleaseNotes_Upcoming.md)

### CODE REVIEW

File: TC/iocBoot/iocTC-IOC-01/st-common.cmd [hunk 1]
```diff
@@ -36,9 +42,10 @@ dbLoadRecords("$(MOTOR)/db/motorUtil.db","P=$(MYPVPREFIX)$(IOCNAME):,$(IFIOC)= ,
 
 ##ISIS## Stuff that needs to be done after all records are loaded but before iocInit is called 
 < $(IOCSTARTUP)/preiocinit.cmd
-
+AdsOpen("$(PORT)", "127.0.0.1", "127.0.0.1.1.1", 500, 500, $(ADS_PORT))
```
[the remark this answers is not in the document]
Comment (rerpha) [reply TC/iocBoot/iocTC-IOC-01/st-common.cmd:45]: ah yes it does, cheers - left that in for testing

File: TC/iocBoot/iocTC-IOC-01/config.xml [hunk 2]
```diff
@@ -3,7 +3,10 @@
 <config_part>
 <ioc_desc>Twincat PV Provider</ioc_desc>
 <macros>
-<macro name="TPY_FILE" pattern="^.+$" description="Name of the .tpy file describing the PLC" hasDefault="NO"/>
+<macro name="IP_AD" pattern="^.+$" description="IP address of controller" hasDefault="YES" defaultValue="192.168.1.221"/>
```
Comment (FreddieAkeroyd) [inline TC/iocBoot/iocTC-IOC-01/config.xml:6]: where is the 192 default applied?

File: TC/iocBoot/iocTC-IOC-01/st-common.lua [hunk 3]
```diff
@@ -1,36 +1,20 @@
...
 	local plc_version = ibex_utils.getMacroValue{macro="PLC_VERSION", default="1"}
-	-- Set TcIoc's scan rate to 150ms for ADS and 300ms for EPICS - this is to avoid ADS fields not updating in channel access quickly enough. 
-	iocsh.tcSetScanRate(150, 2)
-
-	local full_tpy_path = ibex_utils.getMacroValue{macro="TWINCATCONFIG"} .. "/" .. tpy_file
-	if not exists(full_tpy_path) then
-		print("invalid TPY file given: " .. full_tpy_path)
-		iocsh.exit()
-	end
+	local ads_port = ibex_utils.getMacroValue{macro="ADS_PORT", default="852"}
```
Comment (FreddieAkeroyd) [inline TC/iocBoot/iocTC-IOC-01/st-common.lua:11]: do we need "default" here as i believe it is already applied at the st.cmd level?

File: TC/iocBoot/iocTC-IOC-01/st-common.lua [hunk 4]
```diff
@@ -1,36 +1,20 @@
...
+	asyn_port = ibex_utils.getMacroValue{macro="PORT"}
 
+	num_axes = ibex_utils.getMacroValue{macro="NUM_AXES", default="9"} -- todo: actually poll the device to get this
```
Comment (FreddieAkeroyd) [inline TC/iocBoot/iocTC-IOC-01/st-common.lua:14]: config.xml says there is no default for NUM_AXES ?

File: TC/TC_01_settings.req [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0102
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0103
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0104
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0105
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0106
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0107
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0108
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0109
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0110
+file "motor_settings.req" P=TE:NDLT1542:, M=MOT:MTR0111
```
Comment (FreddieAkeroyd) [inline TC/TC_01_settings.req:11]: I wonder is we can simplify this using https://epics-modules.github.io/autosave/autoSaveRestore.html#autosavebuild-automatic-request-file-generation and a `single_axis_tc.req` file - at the moment this is another hard-coded axis limit

Comment (rerpha) [reply TC/TC_01_settings.req:11]: i think that .req file is generated by line 21 and 43 of the lua - perhaps for now i could just unstage that file and add it to the gitignore

**Pull request merged**